cpp/cmake_modules/SetupCxxFlags.cmake

```diff
@@ -48,7 +48,136 @@ if (MSVC)
     set(CXX_COMMON_FLAGS "/W3")
   endif()
 else()
-  set(CXX_COMMON_FLAGS "-Wall -std=c++11")
+  # Common flags set below with warning level
+  set(CXX_COMMON_FLAGS "")
+endif()
+
+# Build warning level (CHECKIN, EVERYTHING, etc.)
+
+# if no build warning level is specified, default to development warning level
+if (NOT BUILD_WARNING_LEVEL)
+  set(BUILD_WARNING_LEVEL Production)
+endif(NOT BUILD_WARNING_LEVEL)
+
+string(TOUPPER ${BUILD_WARNING_LEVEL} UPPERCASE_BUILD_WARNING_LEVEL)
+
+if ("${UPPERCASE_BUILD_WARNING_LEVEL}" STREQUAL "CHECKIN")
+  # Pre-checkin builds
+  if ("${COMPILER_FAMILY}" STREQUAL "msvc")
+    string(REPLACE "/W3" "" CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS}")
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} /W3")
+    # Treat all compiler warnings as errors
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} /WX")
+  elseif ("${COMPILER_FAMILY}" STREQUAL "clang")
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Weverything -Wno-c++98-compat \
+    -Wno-c++98-compat-pedantic -Wno-deprecated -Wno-weak-vtables -Wno-padded \
+    -Wno-unused-parameter -Wno-undef -Wno-documentation-deprecated-sync \
+    -Wno-shadow -Wno-switch-enum -Wno-documentation -Wno-exit-time-destructors \
+    -Wno-global-constructors -Wno-weak-template-vtables -Wno-undefined-reinterpret-cast \
+    -Wno-implicit-fallthrough -Wno-old-style-cast -Wno-unreachable-code-return \
+    -Wno-float-equal -Wno-missing-prototypes -Wno-non-virtual-dtor \
+    -Wno-unused-macros -Wno-covered-switch-default -Wno-unreachable-code-break \
+    -Wno-extra-semi -Wno-cast-align -Wno-shift-sign-overflow \
+    -Wno-used-but-marked-unused -Wno-missing-variable-declarations \
+    -Wno-gnu-zero-variadic-macro-arguments -Wconversion -Wno-sign-conversion \
+    -Wno-disabled-macro-expansion -Wc++11-narrowing -Wnarrowing -Wno-shorten-64-to-32")
```

Comment:
Some of these "pedantic" diagnostics seem useful:

- non-virtual-dtor
- sign-conversion
- unused-parameter (this should probably be a separate patch to clean up unused params)

Reply:
These selections allow the current code to build.

"-Weverything" in Clang enables many more warnings that would otherwise not be seen in Clang or Gcc using other warning flags.

To get the current code to build with fewer warnings suppressed, developers can enable individual diagnostics and remove them from the list when the code is modified (suppress the warning where it occurs if the code does not need to be fixed or change the code so the warning is not issued).